### PR TITLE
Stripe 7: Redesigned product selection and payment

### PR DESCRIPTION
**Pull request by ericpias**

### CODE REVIEW

File: src/Form/index.tsx [hunk 1]
```diff
@@ -913,16 +925,16 @@ function Form({
...
     if (!_draft)
       for (let i = 0; i < activeStep.servar_fields.length; i++) {
-        const servar = activeStep.servar_fields[i].servar;
-        if (servar.type === 'payment_method') {
+        const field = activeStep.servar_fields[i];
+        if (field.servar.type === 'payment_method') {
           const integrationData = integrations?.stripe;
           const actionData: ActionData = {
-            servar,
+            servar: field,
```
Comment (bo-dun-1) [inline src/Form/index.tsx:932]: why is a field being passed as the servar? if we need a field, then we should just call it a field

File: src/Form/index.tsx [hunk 2]
```diff
@@ -564,6 +568,14 @@ function Form({
       updateFieldOptions
     };
 
+    // This could be a redirect from Stripe following a successful payment checkout
+    checkForPaymentCheckoutCompletion(
```
[the remark this answers is not in the document]
Comment (ericpias) [reply src/Form/index.tsx:572]: Ok, moved it up a few lines to where I think you suggested.

File: src/utils/logic.ts [hunk 3]
```diff
@@ -149,16 +149,23 @@ const COMPARISON_FUNCTIONS: {
...
         l,
         r
-      )
-    ),
-  selections_dont_include: (l, r) =>
-    l.every((l: any) =>
+      );
+    return l.some((l: any) => {
+      if (l && detectType(l) === 'object')
+        // Special behavior for left side objects.  Their keys will be compared to the right side
```
[the remark this answers is not in the document]
Comment (ericpias) [reply src/utils/logic.ts:161]: Added this comment:
// While generally useful, this is specifically for the case where the left side
        // is the feathery.payments.selections field which is an object with keys that are the
        // product ids selected for purchase.

File: src/Form/index.tsx [hunk 4]
```diff
@@ -900,6 +914,13 @@ function Form({
       // need to include value === '' so that we can clear out hidden fields
       if (value !== undefined) hiddenFields[fieldKey] = value;
     });
+    // submit feathery reserved hidden fields
+    fieldValues[FEATHERY_PAYMENTS_SELECTIONS] &&
```
Comment (bo-dun-1) [inline src/Form/index.tsx:918]: could we put these into standard if statements? it's a little hard to read in this format

File: src/integrations/stripe.ts [hunk 5]
```diff
@@ -271,6 +338,64 @@ export async function collectPayment(
...
+  if (queryParams.get('_feathery_paid') === 'true') {
+    // remove the _feathery_paid parameter from the url
+    queryParams.delete('_feathery_paid');
+    const { origin, pathname, hash } = featheryWindow().location;
+    const newUrl = `${origin}${pathname}?${queryParams}${hash}`;
+    featheryWindow().history.replaceState({}, '', newUrl);
+
+    // search all step buttons and subgrids for the purchase action and complete the payment
+    const paymentElement = [...step.buttons, ...step.subgrids].find((element) =>
+      element.properties?.actions?.find(
```
Comment (bo-dun-1) [inline src/integrations/stripe.ts:358]: no need for elvis operator on properties (properties?.)

**Review by bo-dun-1 — APPROVED** [review]

**Pull request merged**